src/Storages/MergeTree/MergeTreeIndexText.cpp

```diff
@@ -1383,4 +1495,124 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+
+    expression.execute(block, n_rows);
+
+    return {block.safeGetByPosition(0).column, 0};
+}
+
+String MergeTreeIndexTextPreprocessor::processString(const String &input) const
+{
+    if (expression.getActions().empty())
+        return input;
+
+    Field field(input);
+    ColumnWithTypeAndName entry(column_type->createColumnConst(1, field), column_type, column_name);
+
+    Block block;
+    block.insert(entry);
+
+    size_t nrows = 1;
+    expression.execute(block, nrows);
+
+    return block.safeGetByPosition(0).column->getDataAt(0).toString();
+}
+
+ExpressionActions MergeTreeIndexTextPreprocessor::parseExpression(const IndexDescription & index_description, const String & expression)
+{
+    chassert(index_description.column_names.size() == 1);
+    chassert(index_description.data_types.size() == 1);
+
+    /// Empty expression still creates a preprocessor with empty actions.
+    if (std::ranges::all_of(expression, isspace))
+        return ExpressionActions(ActionsDAG());
+
+    const char * expression_begin = &*expression.begin();
+    const char * expression_end = &*expression.end();
+
+    /// These are expression tokens, do not confuse with index tokens
+    Tokens tokens(expression_begin, expression_end);
+    IParser::Pos token_iterator(tokens, 1000, 1000000);
+
+    Expected expected;
+    ASTPtr expression_ast;
+
+    { /// Parse and verify the expression: String -> ASTPtr
+        if (!ParserExpression().parse(token_iterator, expression_ast, expected))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error parsing preprocessor expression");
+
+        /// Repeat expression validation here. after the string has been parsed into an AST.
+        /// We already made this check during index construction, but "don't trust, verify"
+        const ASTFunction * preprocessor_function = expression_ast->as<ASTFunction>();
+        if (preprocessor_function == nullptr)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Text index preprocessor argument must be an expression");
+
+        // Now we know that the only valid identifier_name must be the text indexed column.
+        String identifier_name = index_description.column_names.front();
+        validatePreprocessorASTExpression(preprocessor_function, identifier_name);
+    }
+
+    /// Convert ASTPtr -> ActionsDAG
+    /// We can do less checks here because in porevious scope we tested that the ASTPtr es an ASTFunction.
+    const String name = expression_ast->getColumnName();
+    const String alias = expression_ast->getAliasOrColumnName();
+
+    NamesAndTypesList source_columns({{index_description.column_names.front(), index_description.data_types.front()}});
+
+    NamesAndTypesList aggregation_keys;
+    ColumnNumbersList aggregation_keys_indexes_list;
+
+    /// Construct a visitor to parse the AST to build a DAG
+    ActionsVisitor::Data visitor_data(
+        Context::getGlobalContextInstance(),
```

Comment:
Recently, I had an issue with the global context that setting changes were not applied. The global context only holds the default values for the settings but does not reflect changes. We need to pass the local context here instead of forwarding the global context.

Reply:
Yes, I understand the struggle to pass local context here (touching many files is pain). We can keep as it is.